XPNet.CLR/Plugin/XPNetPluginBridge.cs

```diff
@@ -506,9 +506,215 @@ internal unsafe struct XPLMCreateFlightLoop_t
          int inRelativeToNow
     );
 
-    #endregion X-Plane Processing API
+	#endregion X-Plane Processing API
+
+	#region X-Plane Display API
+
+	[UnmanagedFunctionPointer(CallingConvention.Cdecl)]
+	internal unsafe delegate int XPLMRegisterDrawCallback(
+		XPLMDrawCallback_f inCallback,
+		XPLMDrawingPhase inPhase,
+		int inWantsBefore,
+		void* inRefcon
+	);
+
+	[UnmanagedFunctionPointer(CallingConvention.Cdecl)]
+	internal unsafe delegate int XPLMUnregisterDrawCallback(
+		XPLMDrawCallback_f inCallback,
+		XPLMDrawingPhase inPhase,
+		int inWantsBefore,
+		void* inRefcon
+	);
 
-    [StructLayout(LayoutKind.Sequential, CharSet = CharSet.Ansi, Pack = 2)]
+	public enum XPLMDrawingPhase : int
+	{
+		xplm_Phase_FirstScene = 0,
+		xplm_Phase_Terrain = 5,
+		xplm_Phase_Airports = 10,
+		xplm_Phase_Vectors = 15,
+		xplm_Phase_Objects = 20,
+		xplm_Phase_Airplanes = 25,
+		xplm_Phase_LastScene = 30,
+		xplm_Phase_FirstCockpit = 35,
+		xplm_Phase_Panel = 40,
+		xplm_Phase_Gauges = 45,
+		xplm_Phase_Window = 50,
+		xplm_Phase_LastCockpit = 55,
+		xplm_Phase_LocalMap3D = 100,
+		xplm_Phase_LocalMap2D = 101,
+		xplm_Phase_LocalMapProfile = 102
+	}
+
+	[UnmanagedFunctionPointer(CallingConvention.Cdecl)]
+	internal unsafe delegate int XPLMDrawCallback_f(
+		XPLMDrawingPhase inPhase,
+		int inIsBefore,
+		void* inRefcon
+	);
+
+	#endregion X-Plane Display API
+
+	#region X-Plane Scenery API
+
+	internal enum XPLMProbeType : int
+	{
+		xplm_ProbeY = 0,
+	}
+
+	public enum XPLMProbeResult : int
+	{
+		xplm_ProbeHitTerrain = 0,
+		xplm_ProbeError = 1,
+		xplm_ProbeMissed = 2
+	}
+
+	[StructLayout(LayoutKind.Sequential, CharSet = CharSet.Ansi, Pack = 2)]
+	internal unsafe struct XPLMProbeInfo_t
+	{
+		public int structSize;
+
+		public float locationX;
+
+		public float locationY;
+
+		public float locationZ;
+
+		public float normalX;
+
+		public float normalY;
+
+		public float normalZ;
+
+		public float velocityX;
+
+		public float velocityY;
+
+		public float velocityZ;
+
+		public int is_wet;
+	};
+
+	[UnmanagedFunctionPointer(CallingConvention.Cdecl)]
+	internal unsafe delegate void* XPLMCreateProbe(
+		XPLMProbeType inProbeType
+	);
+
+	[UnmanagedFunctionPointer(CallingConvention.Cdecl)]
+	internal unsafe delegate void XPLMDestroyProbe(void* inProbe);
+
+	[UnmanagedFunctionPointer(CallingConvention.Cdecl)]
+	internal unsafe delegate XPLMProbeResult XPLMProbeTerrainXYZ(
+								   void* inProbe,
+								   float inX,
+								   float inY,
+								   float inZ,
+								   XPLMProbeInfo_t* outInfo);
+
+	[StructLayout(LayoutKind.Sequential, CharSet = CharSet.Ansi, Pack = 2)]
+	public struct XPLMDrawInfo_t
+	{
+		int structSize;
+
+		float x;
+
+		float y;
+
+		float z;
+
+		float pitch;
+
+		float heading;
+
+		float roll;
+
+		public XPLMDrawInfo_t(float x, float y, float z, float pitch, float heading, float roll)
+		{
+			structSize = System.Runtime.InteropServices.Marshal.SizeOf(typeof(XPLMDrawInfo_t));
```

Comment:
Since the size of this structure can't change at runtime, you should consider getting this in a static constructor so that here you can just assign the known size instead of making a function call to Marshal.SizeOf every time a draw-info object is created.

Reply:
Changed XPLMProbeInfo_t and XPLMDrawInfo_t to be classes instead of structs. The size of the structure is now initialized in the respective static constructor, and copied into the structure at object instantiation.